src/data/reports/glass-data-submission/GLASSDataSubmissionDefaultRepository.ts

```diff
@@ -825,23 +898,143 @@ export class GLASSDataSubmissionDefaultRepository implements GLASSDataSubmission
-        await promiseMap(items, async item => {
-            const programEvents = await this.getProgramEvents(program.id, item.orgUnit ?? "");
+    private async duplicateTrackerProgram(
+        program: ApprovalIds,
+        items: GLASSDataSubmissionItemIdentifier[]
+    ): Promise<void> {
+        const approvedProgramStages = await this.getProgramStages(program.approvedId);
+        const programStages = await this.getProgramStages(program.id);
+
+        _.forEach(items, async item => {
+            const trackedEntities = await this.getTrackedEntities(program.id, item.orgUnit);
+
+            if (!_.isEmpty(trackedEntities)) {
+                const approvedTrackedEntities = await this.getTrackedEntities(program.approvedId, item.orgUnit);
+
+                !_.isEmpty(approvedTrackedEntities) &&
+                    this.api.post(
+                        "/tracker",
+                        { importStrategy: "DELETE" },
+                        { trackedEntities: approvedTrackedEntities }
+                    );
+
+                const trackedEntitiesToPost: TrackedEntity[] = trackedEntities.map(trackedEntity => ({
+                    ...trackedEntity,
+                    trackedEntity: "",
+                    programOwners: trackedEntity.programOwners.map(programOwner => ({
+                        ...programOwner,
+                        program: program.approvedId,
+                    })),
+                    enrollments: trackedEntity.enrollments.map(enrollment => ({
+                        ...enrollment,
+                        enrollment: "",
+                        trackedEntity: "",
+                        program: program.approvedId,
+                        events: enrollment.events.map(event => {
+                            const programStageName = programStages.find(
+                                programStage => programStage.id === event.programStage
+                            )?.name;
+                            const programStageApproved =
+                                approvedProgramStages.find(
+                                    approvedProgramStage => approvedProgramStage.name === programStageName
+                                )?.id ?? "";
+
+                            return {
+                                ...event,
+                                event: "",
+                                trackedEntity: "",
+                                program: program.approvedId,
+                                programStage: programStageApproved,
+                            };
+                        }),
+                    })),
+                }));
+
+                return await this.api.post("/tracker", {}, { trackedEntities: trackedEntitiesToPost }).getData();
+            }
+        });
+    }
+
+    private async getTrackedEntities(program: string, orgUnit: string): Promise<TrackedEntity[]> {
+        let instances: TrackedEntity[] = [];
+        let currentPage = 1;
```

Comment:
although this works, would be better to use the total returned by the api for getting the totalpages. Initialising totalPages to 1 seems hackish :)  what do you think?   

please refer to https://github.com/EyeSeeTea/glass-dev/blob/152a05cb01b44ff4136c04ee64e40ab9e11d0817/src/data/repositories/download-template/DownloadTemplateDefaultRepository.ts#L552

Reply:
oh yes this makes sense. i did not think to use the total returned from the api